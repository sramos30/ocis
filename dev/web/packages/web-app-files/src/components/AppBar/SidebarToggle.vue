<template>
  <oc-button
    id="files-toggle-sidebar"
    v-oc-tooltip="toggleSidebarButtonLabel"
    :aria-label="toggleSidebarButtonLabel"
    appearance="raw"
    class="oc-my-s oc-p-xs"
    @click.stop="toggleSidebar"
  >
    <oc-icon name="side-bar-right" :fill-type="toggleSidebarButtonIconFillType" />
  </oc-button>
</template>

<script>
import { mapState, mapActions } from 'vuex'

export default {
  computed: {
    ...mapState('Files/sidebar', { sidebarClosed: 'closed' }),

    toggleSidebarButtonLabel() {
      if (this.sidebarClosed) return this.$gettext('Open sidebar to view details')
      return this.$gettext('Close sidebar to hide details')
    },

    toggleSidebarButtonIconFillType() {
      return this.sidebarClosed ? 'line' : 'fill'
    }
  },
  methods: {
    ...mapActions('Files/sidebar', { toggleSidebar: 'toggle' })
  }
}
</script>

<style lang="scss" scoped>
#files-toggle-sidebar {
  vertical-align: middle;
  border: 3px solid transparent;
  &:hover {
    background-color: var(--oc-color-background-hover);
    border-radius: 3px;
  }
}
</style>
