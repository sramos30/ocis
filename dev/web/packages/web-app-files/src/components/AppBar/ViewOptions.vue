<template>
  <div>
    <oc-button
      id="files-view-options-btn"
      key="files-view-options-btn"
      v-oc-tooltip="viewOptionsButtonLabel"
      data-testid="files-view-options-btn"
      :aria-label="viewOptionsButtonLabel"
      appearance="raw"
      class="oc-my-s oc-p-xs"
    >
      <oc-icon name="settings-3" fill-type="line" />
    </oc-button>
    <oc-drop
      drop-id="files-view-options-drop"
      toggle="#files-view-options-btn"
      mode="click"
      class="oc-width-auto"
      padding-size="small"
    >
      <oc-list>
        <li class="files-view-options-list-item">
          <oc-switch
            v-model="hiddenFilesShownModel"
            data-testid="files-switch-hidden-files"
            :label="$gettext('Show hidden files')"
          />
        </li>
        <li class="files-view-options-list-item">
          <oc-page-size
            v-model="itemsPerPage"
            data-testid="files-pagination-size"
            :label="$gettext('Items per page')"
            :options="[100, 500, 1000, $gettext('All')]"
            class="files-pagination-size"
          />
        </li>
      </oc-list>
    </oc-drop>
  </div>
</template>

<script>
import { mapMutations, mapState } from 'vuex'
import { useRouteQueryPersisted } from 'web-pkg/src/composables'
import { PaginationConstants } from '../../composables'

export default {
  setup() {
    const perPageQuery = useRouteQueryPersisted({
      name: PaginationConstants.perPageQueryName,
      defaultValue: PaginationConstants.perPageDefault
    })

    return {
      itemsPerPage: perPageQuery
    }
  },
  computed: {
    ...mapState('Files', ['areHiddenFilesShown']),

    viewOptionsButtonLabel() {
      return this.$gettext('Display customization options of the files list')
    },

    hiddenFilesShownModel: {
      get() {
        return this.areHiddenFilesShown
      },

      set(value) {
        this.SET_HIDDEN_FILES_VISIBILITY(value)
      }
    }
  },
  methods: {
    ...mapMutations('Files', ['SET_HIDDEN_FILES_VISIBILITY'])
  }
}
</script>

<style lang="scss" scoped>
#files-view-options-btn {
  vertical-align: middle;
  border: 3px solid transparent;
  &:hover {
    background-color: var(--oc-color-background-hover);
    border-radius: 3px;
  }
}
.files-view-options-list-item {
  & > * {
    display: flex;
    justify-content: space-between;
  }

  & + & {
    margin-top: var(--oc-space-small);
  }
}
</style>
