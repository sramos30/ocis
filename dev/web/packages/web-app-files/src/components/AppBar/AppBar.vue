<template>
  <div id="files-app-bar">
    <oc-hidden-announcer :announcement="selectedResourcesAnnouncement" level="polite" />
    <div class="files-topbar oc-py-s">
      <h1 class="oc-invisible-sr" v-text="pageTitle" />
      <div
        class="oc-flex"
        :class="{
          'oc-flex-between': breadcrumbs.length || hasSharesNavigation,
          'oc-flex-right': !breadcrumbs.length && !hasSharesNavigation
        }"
      >
        <oc-breadcrumb
          v-if="breadcrumbs.length"
          id="files-breadcrumb"
          data-testid="files-breadcrumbs"
          class="oc-flex oc-flex-middle"
          context-menu-padding="small"
          :items="breadcrumbs"
        >
          <template #contextMenu>
            <context-actions
              v-if="showContextActions"
              :items="breadcrumbsContextActionsItems.filter(Boolean)"
            />
          </template>
        </oc-breadcrumb>
        <shares-navigation v-if="hasSharesNavigation" />
        <div v-if="hasViewOptions || hasSidebarToggle" class="oc-flex">
          <view-options v-if="hasViewOptions" />
          <sidebar-toggle v-if="hasSidebarToggle" />
        </div>
      </div>
      <div class="files-app-bar-actions">
        <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px">
          <slot v-if="selectedFiles.length === 0" name="actions" />
          <size-info v-if="showSelectionInfo" class="oc-visible@l" />
          <batch-actions v-if="showBatchActions" />
        </div>
      </div>
      <slot name="content" />
    </div>
  </div>
</template>

<script>
import { mapGetters, mapState, mapMutations } from 'vuex'
import last from 'lodash-es/last'

import Mixins from '../../mixins'
import MixinFileActions from '../../mixins/fileActions'

import BatchActions from './SelectedResources/BatchActions.vue'
import ContextActions from '../FilesList/ContextActions.vue'
import SharesNavigation from './SharesNavigation.vue'
import SizeInfo from './SelectedResources/SizeInfo.vue'
import SidebarToggle from './SidebarToggle.vue'
import ViewOptions from './ViewOptions.vue'

export default {
  components: {
    BatchActions,
    ContextActions,
    SharesNavigation,
    SidebarToggle,
    SizeInfo,
    ViewOptions
  },
  mixins: [Mixins, MixinFileActions],
  props: {
    breadcrumbs: { type: Array, default: () => [] },
    breadcrumbsContextActionsItems: { type: Array, default: () => [] },
    hasBulkActions: { type: Boolean, default: false },
    hasSharesNavigation: { type: Boolean, default: false },
    hasSidebarToggle: { type: Boolean, default: true },
    hasViewOptions: { type: Boolean, default: true }
  },
  computed: {
    ...mapGetters('Files', ['files', 'selectedFiles']),
    ...mapState('Files', ['areHiddenFilesShown']),

    pageTitle() {
      const title = this.$route.meta.title
      return this.$gettext(title)
    },
    areDefaultActionsVisible() {
      return this.selectedFiles.length < 1
    },
    showContextActions() {
      return last(this.breadcrumbs).allowContextActions
    },
    showSelectionInfo() {
      return this.hasBulkActions && this.selectedFiles.length > 0
    },
    showBatchActions() {
      return this.hasBulkActions
    },
    selectedResourcesAnnouncement() {
      if (this.selectedFiles.length === 0) {
        return this.$gettext('No items selected.')
      }
      const translated = this.$ngettext(
        '%{ amount } item selected. Actions are available above the table.',
        '%{ amount } items selected. Actions are available above the table.',
        this.selectedFiles.length
      )
      return this.$gettextInterpolate(translated, { amount: this.selectedFiles.length })
    }
  },

  created() {
    // Storage returns a string so we need to convert it into a boolean
    const areHiddenFilesShown = window.localStorage.getItem('oc_hiddenFilesShown') || 'false'
    const areHiddenFilesShownBoolean = areHiddenFilesShown === 'true'

    if (areHiddenFilesShownBoolean !== this.areHiddenFilesShown) {
      this.SET_HIDDEN_FILES_VISIBILITY(areHiddenFilesShownBoolean)
    }
  },

  methods: {
    ...mapMutations('Files', ['SET_HIDDEN_FILES_VISIBILITY'])
  }
}
</script>

<style lang="scss" scoped>
#files-app-bar {
  background-color: var(--oc-color-background-default);
  border-top-right-radius: 15px;
  box-sizing: border-box;
  z-index: 2;
  position: sticky;
  padding: 0 var(--oc-space-medium);
  top: 0;

  .files-app-bar-actions {
    align-items: center;
    display: flex;
    gap: var(--oc-space-small);
    justify-content: flex-end;
    min-height: 3rem;
  }

  #files-breadcrumb {
    min-height: 2rem;
  }
}
</style>
