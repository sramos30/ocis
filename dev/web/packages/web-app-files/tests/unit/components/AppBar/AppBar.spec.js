import { createLocalVue, shallowMount } from '@vue/test-utils'
import Files from '@/__fixtures__/files'
import Vuex from 'vuex'
import DesignSystem from 'owncloud-design-system'
import GetTextPlugin from 'vue-gettext'

import AppBar from '@files/src/components/AppBar/AppBar.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(DesignSystem)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const selectedFiles = [Files['/'][1], Files['/'][4]]

const actionSlot = "<button class='action-slot'>Click</button>"
const contextMenuSlot = "<button class='context-menu-slot'>Click</button>"
const contentSlot = "<div class='content-slot'>Foo</div>"

const breadcrumbItems = [
  { text: 'Example1', to: '/' },
  { text: 'Example2', to: '/foo' }
]
const breadCrumbItemWithContextActionAllowed = {
  text: 'Example Special',
  to: '/bar',
  allowContextActions: true
}

describe('AppBar component', () => {
  describe('renders', () => {
    it('by default no breadcrumbs, no bulkactions, no sharesnavigation but viewoptions and sidebartoggle', () => {
      const store = createStore()
      const wrapper = getShallowWrapper(store)
      expect(wrapper).toMatchSnapshot()
    })
    describe('breadcrumbs', () => {
      it('if given, by default without breadcrumbsContextActionsItems', () => {
        const store = createStore()
        const wrapper = getShallowWrapper(store, {}, { breadcrumbs: breadcrumbItems })
        expect(wrapper).toMatchSnapshot()
      })
      it('if given, with breadcrumbsContextActionsItems if allowed on last breadcrumb item', () => {
        const store = createStore()
        const wrapper = getShallowWrapper(
          store,
          {},
          { breadcrumbs: [...breadcrumbItems, breadCrumbItemWithContextActionAllowed] }
        )
        expect(wrapper).toMatchSnapshot()
      })
      it('if given, with content in the contextMenu slot', () => {
        const store = createStore()
        const wrapper = getShallowWrapper(
          store,
          { contextMenu: contextMenuSlot },
          { breadcrumbs: [...breadcrumbItems, breadCrumbItemWithContextActionAllowed] }
        )
        expect(wrapper).toMatchSnapshot()
      })
    })
    describe('bulkActions', () => {
      it('if enabled', () => {
        const store = createStore({ selected: selectedFiles })
        const wrapper = getShallowWrapper(store, {}, { hasBulkActions: true })
        expect(wrapper).toMatchSnapshot()
      })
    })
    describe('sharesNavigation', () => {
      it('if enabled', () => {
        const store = createStore()
        const wrapper = getShallowWrapper(store, {}, { hasSharesNavigation: true })
        expect(wrapper).toMatchSnapshot()
      })
    })
    describe('viewoptions and sidebartoggle', () => {
      it('only viewoptions if sidebartoggle is disabled', () => {
        const store = createStore()
        const wrapper = getShallowWrapper(store, {}, { hasSidebarToggle: false })
        expect(wrapper).toMatchSnapshot()
      })
      it('only sidebartoggle if viewoptions is disabled', () => {
        const store = createStore()
        const wrapper = getShallowWrapper(store, {}, { hasViewOptions: false })
        expect(wrapper).toMatchSnapshot()
      })
      it('neither if both are disabled', () => {
        const store = createStore()
        const wrapper = getShallowWrapper(
          store,
          {},
          { hasSidebarToggle: false, hasViewOptions: false }
        )
        expect(wrapper).toMatchSnapshot()
      })
    })
    it('if given, with content in the actions slot', () => {
      const store = createStore()
      const wrapper = getShallowWrapper(store, { actions: actionSlot })
      expect(wrapper).toMatchSnapshot()
    })
    it('if given, with content in the content slot', () => {
      const store = createStore()
      const wrapper = getShallowWrapper(store, { content: contentSlot })
      expect(wrapper).toMatchSnapshot()
    })
  })
})

function getShallowWrapper(
  store = {},
  slots = {},
  props = {
    breadcrumbs: [],
    hasBulkActions: false,
    hasSharesNavigation: false,
    hasSidebarToggle: true,
    hasViewOptions: true
  }
) {
  return shallowMount(AppBar, {
    localVue,
    mocks: {
      $route: {
        meta: {
          title: 'ExampleTitle'
        }
      }
    },
    slots,
    propsData: {
      ...props
    },
    store
  })
}

function createStore(state = { selected: [] }) {
  return new Vuex.Store({
    modules: {
      Files: {
        namespaced: true,
        state,
        getters: {
          selectedFiles: () => state.selected
        },
        mutations: {
          SET_HIDDEN_FILES_VISIBILITY: jest.fn()
        }
      }
    }
  })
}
