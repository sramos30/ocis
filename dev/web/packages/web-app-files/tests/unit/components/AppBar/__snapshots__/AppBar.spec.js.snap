// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar component renders breadcrumbs if given, by default without breadcrumbsContextActionsItems 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-between">
      <oc-breadcrumb-stub items="[object Object],[object Object]" variation="default" contextmenupadding="small" id="files-breadcrumb" data-testid="files-breadcrumbs" class="oc-flex oc-flex-middle"></oc-breadcrumb-stub>
      <!---->
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders breadcrumbs if given, with breadcrumbsContextActionsItems if allowed on last breadcrumb item 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-between">
      <oc-breadcrumb-stub items="[object Object],[object Object],[object Object]" variation="default" contextmenupadding="small" id="files-breadcrumb" data-testid="files-breadcrumbs" class="oc-flex oc-flex-middle"></oc-breadcrumb-stub>
      <!---->
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders breadcrumbs if given, with content in the contextMenu slot 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-between">
      <oc-breadcrumb-stub items="[object Object],[object Object],[object Object]" variation="default" contextmenupadding="small" id="files-breadcrumb" data-testid="files-breadcrumbs" class="oc-flex oc-flex-middle"></oc-breadcrumb-stub>
      <!---->
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders bulkActions if enabled 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="2 items selected. Actions are available above the table."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-right">
      <!---->
      <!---->
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <size-info-stub class="oc-visible@l"></size-info-stub>
        <batch-actions-stub></batch-actions-stub>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders by default no breadcrumbs, no bulkactions, no sharesnavigation but viewoptions and sidebartoggle 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-right">
      <!---->
      <!---->
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders if given, with content in the actions slot 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-right">
      <!---->
      <!---->
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;"><button class="action-slot">Click</button>
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders if given, with content in the content slot 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-right">
      <!---->
      <!---->
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
      </div>
    </div>
    <div class="content-slot">Foo</div>
  </div>
</div>
`;

exports[`AppBar component renders sharesNavigation if enabled 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-between">
      <!---->
      <shares-navigation-stub></shares-navigation-stub>
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders viewoptions and sidebartoggle neither if both are disabled 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-right">
      <!---->
      <!---->
      <!---->
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders viewoptions and sidebartoggle only sidebartoggle if viewoptions is disabled 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-right">
      <!---->
      <!---->
      <div class="oc-flex">
        <!---->
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component renders viewoptions and sidebartoggle only viewoptions if sidebartoggle is disabled 1`] = `
<div id="files-app-bar">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-right">
      <!---->
      <!---->
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <!---->
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;
